libs/ml/include/ml/core/node.hpp

```diff
@@ -141,18 +154,184 @@ class Node
     return static_cast<bool>(cached_output_status_ == CachedOutputState::VALID_CACHE);
   }
 
+  /**
+   * @brief ForwardPassChargeCost calculates charge cost estimation of the current node.
+   * The cost includes also charge costs for every preceding Node in a linear (sequential)
+   * Graph up to input Node.
+   * @return charge cost estimation for all Forward passes through all Nodes starting from input
+   * and ending in this node.
+   */
+  // TODO(VH): Naming does not suggest recursive nature of this method, renaming needed.
+  //           PrecedingNodesForwardCost or just PrecedingForwardCost?
+  virtual fetch::vm::ChargeAmount ForwardPassChargeCost()
+  {
+    // TODO(VH): call OpForwardCost() of the Op - we need to know Data shape to do this.
+    if (input_nodes_.empty())
+    {
+      auto const &expected_in_shapes = op_ptr_->ExpectedSliceInputShapes();
+      if (this->OpCode() == OpType::OP_PLACEHOLDER)
+      {
+        FETCH_LOG_INFO(name_.c_str(), "Input Placeholder layer cost added");
+      }
+      else
+      {
+        FETCH_LOG_ERROR(name_.c_str(),
+                        "Can not estimate a forward pass cost of a non-placeholder leaf Node!");
+      }
+      return op_ptr_->OpForwardCost(expected_in_shapes);
+    }
+    fetch::vm::ChargeAmount total_cost = 0;
+    ShapeVector             input_shapes;
+    for (auto const &i : input_nodes_)
+    {
+      if (auto input_node_ptr = i.lock())
+      {
+        total_cost += input_node_ptr->ForwardPassChargeCost();
+        // TODO(VH): check if the calculations are valid.
+        auto const in_shape = input_node_ptr->SliceOutputShape();
+        if (!in_shape.empty())
+        {
+          input_shapes.emplace_back(in_shape);
+        }
+      }
+      else
+      {
+        throw std::runtime_error("Unable to lock weak pointer.");
+      }
+    }
+    auto const my_cost_of_input = op_ptr_->OpForwardCost(input_shapes);
+    if (std::numeric_limits<fetch::vm::ChargeAmount>::max() - total_cost <= my_cost_of_input)
+    {
+      FETCH_LOG_ERROR(this->name_.c_str(),
```

Comment:
i'm not sure this is actually an error; maybe you just check if an overflow would have occurred and if so set total_cost to max()?

Reply:
I would appreciate a piece of advice on: how to check the overflow here? We have an overflow flag in VM, but this file is not a part of VM itself.